test/rules/avoid_returning_this_to_enable_a_fluent_interface.dart

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N avoid_returning_this_to_enable_a_fluent_interface`
+
+class A {
+  int x;
+  A badAddOne() {
+    x++;
+    return this; // LINT
+  }
+
+  int goodAddOne() {
+    x++;
+    return this.x; // OK
+  }
+  A getThing() {
+    return this; // OK
+  }
+}
+
+class B extends A{
+  @override
+  badAddOne() { // It is ok because it is an inherited method.
```

Comment:
Perhaps add a return type of "B" so we know that the reason there's no lint is because it overrides an inherited method and not because the type is not the same as the declaring class. (And add another test that shows that we don't create a lint if a method has a different return type even when it doesn't override an inherited method.)